core/src/avm2/activation.rs

```diff
@@ -97,14 +98,19 @@ pub struct Activation<'a, 'gc: 'a, 'gc_context: 'a> {
     #[allow(dead_code)]
     return_value: Option<Value<'gc>>,
 
-    /// The current local scope, implemented as a bare object.
-    #[allow(dead_code)]
-    local_scope: Object<'gc>,
-
     /// The current scope stack.
+    scope_stack: ScopeStack<'gc>,
+
+    /// This represents the outer scope of the method that is executing.
+    /// The outer scope gives an activation access to the "outer world", including
+    /// the current Domain.
+    outer: ScopeChain<'gc>,
+
+    /// The context of the method that is executing.
+    /// This is used exclusively by builtin methods to gain access to the Domain of the caller.
     ///
-    /// A `scope` of `None` indicates that the scope stack is empty.
-    scope: Option<GcCell<'gc, Scope<'gc>>>,
+    /// This will always be available to builtin methods, so it is safe to unwrap.
+    code_context: Option<Domain<'gc>>,
```

Comment:
Is there any particular reason in which we would want to store `None` here, as opposed to, say, falling back to the global domain if this activation was created for something that does not have an application domain?

Reply:
Also, nitpick: I find the name `code_context` to be less descriptive than it could be. I understand that this name was taken from avmplus; I'm not sure if the name makes sense here. In Ruffle, a "context" is a group of related borrows intended to grant access to various parts of the same player for code that needs to be called at specific times. (e.g. we have `UpdateContext` for code that's called when updating player state, and `RenderContext` for code that renders it. At one point we even had an `ActionContext`.)

Maybe something like `caller_domain` might work? The name immediately explains what's up: it's the domain of the thing that called the function in the current activation.

Reply:
So one of the reasons I thought making it an Option was a good idea was so that if a builtin method calls another builtin method, the caller domain will still be the domain of the original AS3 caller, and not the domain of the builtin method that called it. In order to accomplish that, we need some way to know if the activation we are calling from is from a builtin or not. 

Maybe instead we could have a boolean in the activation that indicates if it is from a builtin method or not?